docs/usage/WritingCustomMiddleware.md

```diff
@@ -0,0 +1,146 @@
+  beforeSideEffect(action, beforeState, api)
+  return next(action)
+}
+```
+
+### Late side effect
+
+```ts
+const middleware: Middleware = api => next => action => {
+  const response = next(action)
+
+  // Do something after the action hits the reducer
+  const afterState = api.getState()
+  afterSideEffect(action, afterState, api)
+
+  return response
+}
+```
+
+## Rules to make compatible middleware
+
+In principle, middleware is a very powerful pattern and can do whatever it wants with an action. Existing middleware might have assumptions about what happens in the middleware around it, though, and being aware of these assumptions will make it easier to ensure that your middleware works well with existing commonly used middleware.
+
+There are two contact points between our middleware and the other middlewares:
+
+### Calling the next middleware
+
+When you call `next`, the middleware will expect some form of action. Unless you want to explicitly modify it, just pass through the action that you received.
+
+More subtly, some middlewares except that the middleware is called on the same tick as `dispatch` is called, so next should be called synchronously by your middleware.
+
+### Returning the dispatch return value
+
+Unless the middleware needs to explicitly modify the return value of `dispatch`, just return what you get from `next`. It you do need to modify the return value, then your middleware anyway will need to sit in a very specific spot in the middleware chain to be able to do what it is supposed to, so you will need to check compatability with all other middlewares manually and decide how they could work together.
+
+This has a tricky consequence:
+
+```ts
+const middleware: Middleware = api => next => async action => {
+  const response = next(action)
+
+  // Do something after the action hits the reducer
+  const afterState = api.getState()
+  if (action.type === 'REQUEST_FETCH') {
+    const data = await fetchData()
+    api.dispatch(dataFetchedAction(data))
+  }
+
+  return response
+}
+```
+
+Even though it looks like we didn't modify the response, we actually did: Due to async-await, it is now a promise. This will break some middlewares like e.g. the one from RTK query.
+
+So how to write this middleware?
+
+```ts
+const middleware: Middleware = api => next => action => {
+  const response = next(action)
+
+  // Do something after the action hits the reducer
+  const afterState = api.getState()
+  if (action.type === 'REQUEST_FETCH') {
+    void loadData(api)
+  }
+
+  return response
+}
+
+async function loadData(api) {
+  const data = await fetchData()
+  api.dispatch(dataFetchedAction(data))
+}
+```
+
+Just move out the async logic into a separate function so that you can still use async-await, but don't actually wait for the promise to resolve in the middleware. `void` indicates to others reading the code that you decided to not await the promise explicitly without having an effect on the code.
```

Comment:
:speech_balloon: hmm. tbh I've never seen that `void` pattern used before 🤷‍♂️

Reply:
It's recommended e.g. by [this eslint rule](https://typescript-eslint.io/rules/no-floating-promises/) that is part of eslint's configuration set [`recommended-requiring-type-checking`](https://typescript-eslint.io/linting/configs/#recommended-requiring-type-checking). I've also seen it in other codebases as a nice convention.